src/sentry/ingest/ingest_consumer.py

```diff
@@ -53,16 +58,37 @@ def process_message(self, message):
             )
             return  # message already processed do not reprocess
 
-        cache_key = cache_key_from_project_id_and_event_id(project_id=project_id, event_id=event_id)
+        try:
+            project = Project.objects.get_from_cache(id=project_id)
+        except Project.DoesNotExist:
+            logger.error("Project for ingested event does not exist: %s", project_id)
+            return
+
+        # Parse the JSON payload. This is required to compute the cache key and
+        # call process_event. The payload will be put into Kafka raw, to avoid
+        # serializing it again.
+        # XXX: Do not use CanonicalKeyDict here. This may break preprocess_event
+        # which assumes that data passed in is a raw dictionary.
+        data = json.loads(payload)
+
         cache_timeout = 3600
-        default_cache.set(cache_key, body, cache_timeout, raw=True)
+        cache_key = cache_key_for_event(data)
+        default_cache.set(cache_key, payload, cache_timeout, raw=True)
 
-        # queue the event for processing
-        preprocess_event.delay(cache_key=cache_key, start_time=start_time, event_id=event_id)
+        # Preprocess this event, which spawns either process_event or
+        # save_event. Pass data explicitly to avoid fetching it again from the
+        # cache.
+        preprocess_event(cache_key=cache_key, data=data, start_time=start_time, event_id=event_id)
 
         # remember for an 1 hour that we saved this event (deduplication protection)
         cache.set(deduplication_key, "", 3600)
 
+        # emit event_accepted once everything is done
+        canonical_data = CanonicalKeyDict(data)
```

Comment:
This uses `CanonicalKeyDict`, but it could actually use `CanonicalKeyView`, which is substantially faster. It depends on whether we assume that users can modify the dict passed in or not. 
@tonyo  what do you think about this?